### DISCUSSION TITLE
📣 Are you using `when` expressions?  📣

### DISCUSSION THREAD

**Discussion by jerop** [tektoncd/pipeline#4185] [General]
Hello Tekton Friends!

This discussion is about changes to `when` expressions in [Tekton Pipelines 0.27](https://github.com/tektoncd/pipeline/releases/tag/v0.27.0).

We are migrating to using `when` expressions to guard a `Task` only. For further details, read the proposal in [TEP-0059: Skipping Strategies](https://github.com/tektoncd/community/blob/main/teps/0059-skipping-strategies.md). 

Set `scope-when-expressions-to-task` field in [feature flags](https://github.com/tektoncd/pipeline/blob/v0.27.1/docs/install.md#customizing-the-pipelines-controller-behavior) to `"true"` enable the new functionality. 

#### Guard a `Task` only

When guarding a `Task` only, its:
- ordering-dependent `Tasks`, based on `runAfter`, will be executed
- resource-dependent `Tasks` will be attempted and:
   - if based on `Results`, the dependent `Tasks` will be skipped due to missing `Results`
      - further dependent `Tasks` will be skipped as well
      - when we support default `Results`, these resource-dependent `Tasks` would be executed if default `Results` are specified, read more in [TEP-0048](https://github.com/tektoncd/community/blob/main/teps/0048-task-results-without-results.md)
   - if based on `Workspaces`, make sure to handle the execution of the dependent `Task` in case the file expected from the skipped parent `Task` is missing from the `Workspace`

Read more in [Guard a `Task` only](https://github.com/tektoncd/pipeline/blob/v0.27.1/docs/pipelines.md#guarding-a-task-only).

#### Guard a `Task` and its dependent `Tasks`
To guard a `Task` and its dependent `Tasks`, there are two options:
- cascade the `when` expressions to the specific dependent `Tasks` to be guarded as well
- compose the `Task` and its dependent `Tasks` as a unit to be guarded and executed together using `Pipelines in Pipelines` (experimental feature), read more in [TEP-0056](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md)

Read more in[ Guard a `Task` and its dependent `Tasks`](https://github.com/tektoncd/pipeline/blob/v0.27.1/docs/pipelines.md#guarding-a-task-and-its-dependent-tasks).

#### Example

In the example below, _manual-approval_ `Task` is executed only when merging a pull request, which is guarded using `when` expressions. To reuse the `Pipeline` when merging and not merging, the dependent `Tasks` need to be executed when the guarded _manual-approval_ `Task` is skipped.

<img width="268" alt="Screen Shot 2021-08-17 at 12 19 24 PM" src="https://user-images.githubusercontent.com/18076400/130492966-00fe31dc-895e-4aa9-8ba4-4a300ff06765.png">

Previously, if the `when` expressions in _manual-approval_ were evaluated to `"false"`, then _manual-approval_, _build-image_, _deploy-image_ and _slack-msg_ `Tasks` would all be skipped.

With scoping `when` expressions to `Task`, then _build-image_ and _deploy-image_ would be executed. For now, _slack-msg_ would be skipped because of the missing approver `Result `- but should be executed if you provide default `Results` when we support them. 

#### Timeline
In February 2022, this flag will be flipped to `"true"` and then it’ll be removed in March 2022. Read more in [Deprecations](https://github.com/tektoncd/pipeline/blob/main/docs/deprecations.md) and [TEP-0059: Skipping Strategies - Migration](https://github.com/tektoncd/community/blob/main/teps/0059-skipping-strategies.md#upgrade--migration-strategy).  

Happy to answer any questions here or elsewhere. I’ll follow up with updates and reminders on this migration. 

Thanks, 
Jerop

**Comment by afrittoli** [comment]
The flag default value was switched as part of release v0.33.x
